### PR TITLE
Refine file naming scheme for profiling data

### LINKED ISSUE #336: Incorrect basename generated by dynamic profiler
How to reproduce?
1. `git clone https://github.com/sysprog21/rv32emu.git --depth=1`
2. `cd rv32emu`
3. `make`
4. `mkdir prof`
5. `build/rv32emu -p build/hello.elf`
6. `ls prof`

You will see:
```
'llo'$'\206\025\177''.prof' 
```

Reason:
"main.c"
```
if (opt_prof_data) {
        char *prog_name = malloc(strlen(opt_prog_name) - 11);
        strncpy(prog_name, opt_prog_name + 8, strlen(opt_prog_name) - 12);
        prof_out_file = malloc(strlen(opt_prog_name) + 1);
        sprintf(prof_out_file, "./prof/%s.prof", prog_name);
    }
```

I think the above codes want to get the basename of the target program. But, it is only valid  if target program is specified as `./build/xxx.elf`. 

Solution:
Maybe using `man 3 basename` is easier and sufficient.

### PR DESCRIPTION
**Pull request by ChinYikMing**
Close #336

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -167,10 +169,10 @@ static bool parse_args(int argc, char **args)
...
     opt_prog_name = prog_args[0];
     if (opt_prof_data) {
-        char *prog_name = malloc(strlen(opt_prog_name) - 11);
-        strncpy(prog_name, opt_prog_name + 8, strlen(opt_prog_name) - 12);
-        prof_out_file = malloc(strlen(opt_prog_name) + 1);
-        sprintf(prof_out_file, "./prof/%s.prof", prog_name);
+        char *prog_basename = basename(opt_prog_name);
+        prof_out_file = malloc(strlen(prog_basename) + 12 + 1);
+        assert(prof_out_file);
+        sprintf(prof_out_file, "./prof/%s.prof", prog_basename);
```
Comment (jserv) [inline src/main.c]: The leading `prof/` directory name is not really making sense. Instead, let's place the profiling data in the same directory where the RV32 ELF file locates.

Comment (ChinYikMing) [reply src/main.c]: I think you mean `build` directory

**Comment by ChinYikMing** [comment]
Note that the trailing spaces are removed automatically by editor when I motify "README" and "rv_profiler", not my primary intention.

**Review by qwe661234 — APPROVED** [review]

File: src/main.c (hunk 1 above)
Comment (qwe661234) [reply src/main.c]: I think putting it in the `build` directory is reasonable, we can distinguish the executable files and profiling data by filename extension.

File: src/main.c [hunk 2]
```diff
@@ -167,10 +169,10 @@ static bool parse_args(int argc, char **args)
...
     opt_prog_name = prog_args[0];
     if (opt_prof_data) {
-        char *prog_name = malloc(strlen(opt_prog_name) - 11);
-        strncpy(prog_name, opt_prog_name + 8, strlen(opt_prog_name) - 12);
-        prof_out_file = malloc(strlen(opt_prog_name) + 1);
-        sprintf(prof_out_file, "./prof/%s.prof", prog_name);
+        char *prog_basename = basename(opt_prog_name);
+        prof_out_file = malloc(strlen(prog_basename) + 11 + 1);
+        assert(prof_out_file);
+        sprintf(prof_out_file, "build/%s.prof", prog_basename);
```
Comment (jserv) [inline src/main.c]: Is the leading `build/` necessary?

Comment (ChinYikMing) [reply src/main.c]: It depends on working directory of `rv32emu` executable. There are 2 cases to run profiler:

1. 
```
// outside build directory
xxx/build/rv32emu -p xxx/build/xxx.elf
```

2. 
```
// inside build directory
./rv32emu -p xxx.elf
```

Without leading `build/`, the first case will always write to cwd which is parent directory of `xxx`. For second case, it is all right.

I think the new approach could be concatenating the cwd and relative path (e.g., `xxx/build/` in first case and `./` in second case) to `rv32emu` executable. This concatenated path will always point to the `build` directory.

Comment (jserv) [reply src/main.c]: > I think the new approach could be concatenating the cwd and relative path (e.g., `xxx/build/` in first case and `./` in second case) to `rv32emu` executable. This concatenated path will always point to the `build` directory.

I think the new one is better, and I expect a cleaner and elegant implementation.

**Comment by jserv** [comment]
Thank @ChinYikMing for contributing!

**Pull request merged**